File: elixir/day3_part2_2025.exs

defmodule Main do
  @k 12

  # entry point ---------------------------------------------------------------
  def main do
    total =
      "input.txt"
      |> File.stream!()
      |> Stream.map(&String.trim/1)
      |> Stream.filter(&valid_line?/1)
      |> Enum.reduce(0, fn line, acc ->
        best = max_subseq(line, @k)
        acc + String.to_integer(best)
      end)

    IO.puts("Total output joltage: #{total}")
  end

  # ---------------------------------------------------------------------------
  defp valid_line?(s), do: s != "" and String.length(s) >= @k

  # ---------------------------------------------------------------------------
  # Returns the maximum numeric subsequence of length k from the string s.
  # Implements the monotonic‑stack algorithm in O(n) time.
  defp max_subseq(s, k) do
    to_remove = String.length(s) - k

    {stack, _} =
      s
      |> String.graphemes()
      |> Enum.reduce({[], to_remove}, fn d, {stk, rem} ->
        {new_stk, new_rem} = pop_while(stk, rem, d)
        {[d | new_stk], new_rem}
      end)

    stack
    |> Enum.reverse()
    |> Enum.take(k)
    |> Enum.join()
  end

  # Pops from the stack while the top element is smaller than the current digit
  # and we still have removals left.
  defp pop_while([h | t] = stk, rem, d) when rem > 0 and h < d do
    pop_while(t, rem - 1, d)
  end

  defp pop_while(stk, rem, _d), do: {stk, rem}
end

# Run the program -------------------------------------------------------------
Main.main()

File: fsharp/day3_part2_2025.fs

open System
open System.IO

let findLargest (s:string) k =
    let n = s.Length
    let mutable toRemove = n - k
    let stack = Array.zeroCreate n
    let mutable top = -1
    for i in 0 .. n - 1 do
        let c = s.[i]
        while toRemove > 0 && top >= 0 && stack.[top] < c do
            top <- top - 1; toRemove <- toRemove - 1
        top <- top + 1; stack.[top] <- c
    String(stack, 0, k)

[<EntryPoint>]
let main _ =
    File.ReadLines("input.txt")
    |> Seq.sumBy (fun line ->
        let trimmed = line.Trim()
        if trimmed.Length = 0 then 0L
        else int64 (findLargest trimmed 12))
    |> printfn "%d"
    0
